Adding the process of adding/updating the icons

iifawzi opened the pull request:
Hi, it's been a while, but it's finally done. 
I will make a reviews to elaborate more on some decisions
![c875fcfb-4615-424b-9058-e699ac650a3e](https://user-images.githubusercontent.com/46695441/124369984-37792200-dc72-11eb-9b26-692e51b7d38f.JPG)

The whole process of adding 1788 icon is taking around 11/12 seconds, most of them taken by `grunt` and `npm` to update the package and move the files.

iifawzi commented on package.json:
```diff
@@ -8,10 +8,11 @@
   },
   "scripts": {
     "preinstall": "npm update eos-icons ",
-    "postinstall": "grunt -b ./src moveFiles && cp node_modules/eos-icons/animated-svg/*.svg ./src/svg/",
```
The part of moving `animated-svg` icons, is added to grunt, since all coping process is already being made there.

iifawzi commented on src/components/icons/interfaces.icons.ts:
```diff
@@ -1,16 +1,18 @@
 
 // Interface of the icons objects stored in db:
-export interface IconInterface {
+export interface IconInterface extends mongoose.Document{
```
extended to add the `_id` field

iifawzi commented on src/helpers/Logger.ts:
```diff
@@ -6,30 +6,40 @@ const dateFormat = () => {
+    this.logger.error(err.message, { source, logDetails: err })
+  }
+
+  logInfo (source: string, info: InfoLog) {
+    this.logger.info(info.message, { source, logDetails: info })
```
info logger is added to be used when we need to log an `info` logs, it will be saved to `info.log` and is used in this PR to an info while

iifawzi commented on src/utils/tools/updatingIcons/getEncodedLink.ts:
```diff
@@ -0,0 +1,16 @@
+
+/**
+ *
+ * Generate encoded links, because gitlab is using an encoded file paths
+ *
+ */
+const getEncodedLink = (pathToFile: string) => {
+  const APILink = configs.GITLAB_READ_API
+  const encodedPath = pathToFile.replace(/\//g, '%2F')
+  const options = 'raw?ref=master'
+  const encodedLink = `${APILink}/${encodedPath}/${options}`
+  return encodedLink
```
This used to replace each `/` with `%2F` for gitlab to be able to locate the files

iifawzi commented on src/utils/tools/updatingIcons/getSvgCode.ts:
```diff
@@ -0,0 +1,28 @@
+import { iconsTheme } from 'common/types'
+import fs from 'fs'
+
+const getFilled = async (iconName: string) => {
+  const iconPath = `${ThemesDirectories[iconsTheme.filled]}/${iconName}.svg`
+  if (fs.existsSync(iconPath)) {
+    const svgCode = fs.readFileSync(iconPath, 'utf8')
+    return svgCode
+  } else {
+    const outlinedSvgCode = await getOutlined(iconName)
+    return outlinedSvgCode
+  }
+}
```
For the `filled` and `animated` icons, It will look for it first in the `svg` folder, if it's not available, it will look in `svg-outlined` folder , since the material icons are combined to their main directories, it enough to check the main directories.

iifawzi commented on src/utils/tools/updatingIcons/isNewIcon.ts:
```diff
@@ -0,0 +1,8 @@
+const isNewIcon = (iconCreationDate: string, currentVersionDate: Date) => {
+  // icon creation date's style is: DD/MM/yyyy, we need to convert it to the standard: yyyy/DD/MM
+  // to be able to make a comparison.
+  const creationDate = new Date(iconCreationDate.split('/').reverse().join('-'))
+  return creationDate > currentVersionDate
+}
+
+export default isNewIcon
```
This function is used to convert the date of the icons which's stored in `eos-icons.json` file to the standard, this method should be removed in the future, if `eos-icons.json` used the standard `yyyy-DD-MM`

iifawzi commented on src/utils/tools/updatingIcons/prepareTheUpdate.ts:
```diff
@@ -0,0 +1,28 @@
+
+const prepareUpdatedIcon = (newIconDetails: IconInterface) => {
+  const updateDetails = {
+    $set: {},
+    $addToSet: {}
```
`$set` is used for full replacement, this will be used to use all the incoming fields except, `tags`, `categories` since they're an arrays.
`$addToSet` is used for the `tags`, and  `category`, to add the tags or categories that aren't already exist.

iifawzi commented on src/utils/tools/updatingIcons/prepareTheUpdate.ts:
```diff
@@ -0,0 +1,28 @@
+    if (key === 'tags') {
+      updateDetails.$addToSet[key] = { $each: value }
+    }
+    // category might be a string or an array
```
Category is sometimes string, and sometimes an array, this check can be omitted, if `eos-icons.json` used a fixed schema for all the icons.

iifawzi commented on src/utils/updateDBIcons.ts:
```diff
@@ -0,0 +1,129 @@
+    const currentVersionInfo = await infoServices.getLatestVersionInfo()
+    // in initial setup, there wouldn't be any versions stored in the db, that's why we set an old date.
+    const currVersionDate = new Date(currentVersionInfo ? currentVersionInfo.createdAt : '1999-03-20')
+    if (!notifiedByHook) {
+      // if not notified by hook, we need to make sure that there's a new release first
+      const eosIconsTAGS = await axios.get(configs.GITLAB_TAGS_API)
+      const latestReleaseDate = eosIconsTAGS.data[0].commit.committed_date
+      if (latestReleaseDate <= currVersionDate) {
+        updateIconsLogger.logInfo('', { message: 'icons is up to date, updating process is finished' })
+        return
+      }
+    }
```
In most cases the update icons process will be called without being notified by the gitlab hook (will be created in the next PR), so we need to make sure that there's a new version of icons before starting the process, if a hook notified our backend that there's a new release, this check will not be made.

iifawzi commented on src/utils/updateDBIcons.ts:
```diff
@@ -0,0 +1,129 @@
+    // Latest Database version's info:
+    const currentVersionInfo = await infoServices.getLatestVersionInfo()
+    // in initial setup, there wouldn't be any versions stored in the db, that's why we set an old date.
+    const currVersionDate = new Date(currentVersionInfo ? currentVersionInfo.createdAt : '1999-03-20')
```
This's a random `date` to be used when there's no versions at all in the `db`, to be able to fetch all the icons, in the initialization of the db phase. 

it's not random tbh, it's my bd :p. Anyway, it can be any old date.

kaybhutani replied:
Why not use the current date to know when it was populated?

iifawzi replied:
I don't know how I missed this comment! sorry for the late reply @kbhutani0001. 

if I understood your question correctly, you mean using `new Date()` instead of `1999-03-20` right? 
We will only use `1999-03-20` in case, if there's no info records in the db (server is initially configured), it means that we must fetch all the icons, regardless of its released date, that's why I've used a very old date `1999-03-20`, by using this old date, this condition will return `false`, and the icons will be added. 

```JS
    const currentVersionInfo = await infoServices.getLatestVersionInfo()
    const currVersionDate = new Date(currentVersionInfo ? currentVersionInfo.createdAt : '1999-03-20')
    if (latestReleaseDate <= currVersionDate) {
        return
    }
```

in other hand, if the icons are configured before, it means that there will be an info records in the db, in the next attempt to update the icons, we will check whether the `latest released package` has a release date lower or larger than the latest info about our icons (from the info's record), if the release date is bigger, it means that there's a new released tag.. if it's lower or equal, it means that there's no new releases, so we don't need to update the icons and will return. 

The question that might arise is, isn't the webhook will always notify our backend when a new release is released? Totally right, i've only made this check because this update process can be called manually, and actually, it will be called each time the server is booting up, so I needed to make sure that I don't update the icons unless there's a new version.

kaybhutani replied:
got it. makes sense.

iifawzi commented on src/utils/updateDBIcons.ts:
```diff
@@ -0,0 +1,129 @@
+      // ADD the SVG codes to the icon's object:
+      const filledSVG = await getFilled(iconDetails.name)
+      if (filledSVG === 'NOT FOUND') {
+        continue
```
if no code is found for any icon, it will be skipped.

iifawzi commented on src/utils/updateDBIcons.ts:
```diff
@@ -0,0 +1,129 @@
+          // if not found, use the same svg code of the filled version
+          iconDetails.svgOutlined = iconDetails.svg
+        } else {
+          iconDetails.svgOutlined = outlinedSVG
```
As per our discussions in slack, if an icon should have an outlined version, and we couldn't find the file, we will use the same `svg` code.

iifawzi reviewed (commented):
Those are the things that I wanted to mention, I've added as most comments as I can to make it clear for future contributors for ourselves as well.

iifawzi commented on package.json:
```diff
@@ -8,13 +8,13 @@
     "build": "tsc --project ./",
-    "start:dev": "grunt -b ./src moveFiles && NODE_PATH=./src nodemon ./src/server.ts",
-    "start": "grunt -b ./src moveFiles && NODE_PATH=./src ts-node --transpile-only ./src/server.ts",
+    "start:dev": "NODE_PATH=./src nodemon ./src/server.ts",
+    "start": "NODE_PATH=./src ts-node --transpile-only ./src/server.ts",
     "lint": "eslint \"src/*\"",
     "lint:fix": "eslint \"src/*\" --fix",
-    "test:v1": "grunt -b ./src moveFiles && NODE_ENV=test NODE_PATH=./src ts-node --transpile-only ./src/server.ts"
```
I've removed the `moveFiles` call from the scripts, since it will be called from inside the update process.

iifawzi replied:
I've kept the moveFiles call only for the `test` environment, since the update process will not be fired in the test environments, now all tests are passing.

iifawzi commented on src/app.ts:
```diff
@@ -20,7 +21,8 @@ const app:Express.Application = Express();
 
   // ************** Other App configurations *****************
-
+  // Run the update Icons Process:
+  updateDBIcons()
```
This will always run when the server starts

iifawzi commented:
Since we're planning to use the `version 1 ` APIs, this's the `updateDBicons.ts` method that will be used in production.  
I've just removed the database part from the code that will be deployed, since the `version 1` apis will not deal with the db's icons. 

You can take a look at the code that will deployed in the `heroku-stable-version` branch. 

```JS
const updateDBIcons = async () => {
  try {
    console.time('Updating the icons Process')
    updateIconsLogger.logInfo('', { message: 'The process of updating the icons is started' })
    // Update the local eos-icons package and move the files using grunt:
    await pcmdGet('npm update eos-icons')
    await pcmdGet('grunt -b ./src moveFiles')
    updateIconsLogger.logInfo('Updating the local package, and moving the files', { message: 'Finished' })
    updateIconsLogger.logInfo('', { message: 'The process of updating the icons is Finished' })
    console.timeEnd('Updating the icons Process')
  } catch (err) {
    NodeLogger.logError('UdpateDBIcons', err)
  }
}
```

No magic, when our server is notified, it will update the local dependency and use grunt to prepare the files.
I've added these lines to give us more information about the process, if you think that this's unnecessary I will delete them.

kaybhutani commented on src/utils/updateDBIcons.ts:
```diff
@@ -0,0 +1,115 @@
+      }
+    }
+
+    if (newIcons.length !== 0) {
```
```suggestion
    if (!!newIcons.length) {
```
